flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,811 @@
+ *       not implemented yet.
+ * </ul>
+ *
+ * The job graph looks like below:
+ *
+ * <pre>{@code
+ *                            Flink sink
+ *               +-----------------------------------------------------------------------------------+
+ *               |                                                                                   |
+ * +-------+     | +----------+                               +-------------+      +---------------+ |
+ * | Map 1 | ==> | | writer 1 |                               | committer 1 | ---> | post commit 1 | |
+ * +-------+     | +----------+                               +-------------+      +---------------+ |
+ *               |             \                             /                \                      |
+ *               |              \                           /                  \                     |
+ *               |               \                         /                    \                    |
+ * +-------+     | +----------+   \ +-------------------+ /   +-------------+    \ +---------------+ |
+ * | Map 2 | ==> | | writer 2 | --->| commit aggregator |     | committer 2 |      | post commit 2 | |
+ * +-------+     | +----------+     +-------------------+     +-------------+      +---------------+ |
+ *               |                                             Commit only on                        |
+ *               |                                             committer 1                           |
+ *               +-----------------------------------------------------------------------------------+
+ * }</pre>
+ */
+@Experimental
+public class IcebergSink
+    implements Sink<RowData>,
+        SupportsPreWriteTopology<RowData>,
+        SupportsCommitter<IcebergCommittable>,
+        SupportsPreCommitTopology<WriteResult, IcebergCommittable>,
+        SupportsPostCommitTopology<IcebergCommittable> {
+  private final Table initTable;
+  private final TableLoader tableLoader;
+  private final Map<String, String> snapshotProperties;
+  private final String uidPrefix;
+  private final String sinkId;
+  private final Map<String, String> writeProperties;
+  private final RowType flinkRowType;
+  private final SerializableSupplier<Table> tableSupplier;
+  private final transient FlinkWriteConf flinkWriteConf;
+  private final List<Integer> equalityFieldIds;
+  private final boolean upsertMode;
+  private final FileFormat dataFileFormat;
+  private final long targetDataFileSize;
+  private final String branch;
+  private final boolean overwriteMode;
+  private final int workerPoolSize;
+  private final Table table;
+  private final List<String> equalityFieldColumns = null;
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSink.class);
+
+  private IcebergSink(
+      Table initTable,
+      TableLoader tableLoader,
+      Table table,
+      Map<String, String> snapshotProperties,
+      String uidPrefix,
```

Comment:
I think we will create our own operators. Not just Writer/Committer/GlobalCommitter, but the operator in `PreCommitToplogy` and the operators in the `PostCommitTopology` as well. I think we would want to stick to the `uidPrefix` constrcut

Reply:
how do we set uid for the operators created by the v2 sink framework like writer and committer operators?

Reply:
I haven't been able to find a way to propagate the `uidPrefix` from the `IcebergSink` to the operators created by the v2 sink framework. There's this `CustomSinkOperatorUidHashes` where one could maybe plug in some uid hashes, but it's not clear to me how to generate this hashes, and if this is even the right approach. It seems to me that having to hash the uidPrefix is not the right approach? But I maybe wrong here. 

So, something like this could work:
```java
public DataStreamSink<RowData> append() {
      IcebergSink sink = build();
      DataStream<RowData> rowDataInput = inputCreator.apply(uidPrefix);
      String writeUidHash = hashFunction(uidPrefix+"-writer");
      String commiterUidHash = hashFunction(uidPrefix+"-committer");
      String globalCommitterUiHash = hashFunction(uidPrefix+"-global-committer");
      CustomSinkOperatorUidHashes hashes = new CustomSinkOperatorUidHashes(writeUidHash, commiterUidHash, globalCommitterUiHash);
      DataStreamSink<RowData> rowDataDataStreamSink =
          rowDataInput.sinkTo(sink, hashes).uid(uidPrefix + "-sink");
      if (sink.flinkWriteConf.writeParallelism() != null) {
        rowDataDataStreamSink.setParallelism(sink.flinkWriteConf.writeParallelism());
      }
      return rowDataDataStreamSink;
    }
```
`hashFunction` being the hashing function (I'm not sure which one should be used here)

to me, the above seems a little hacky. There should be a more straightforward way to set this up.

Reply:
previously, we control the creation of all operators and apply `uidPrefix` consistently across all operators. 

with v2 sink framework, some operators (like aggregator) are created by the Iceberg connector and some operators (like writer and committer) are created by the framework. The committer operator is a stateful operator. Does the framework ensure a deterministic `uid` for the committer operator or is it a random UUID as the operator id?

Reply:
I did a debugging session, where I could verify that the framework created operators were prefixing a value to the iceberg sink uid. In this example, the sink uid was "meneses" and the below picture was for the `Sink Committer` operator
![image](https://github.com/user-attachments/assets/ce57db19-de46-4cca-a700-b4a1cda8b628)

In this way, the Sink UID is actually being propagated to all of these operators, which is what we wanted.